File: org.dcs.nifi.processors/src/main/java/org/dcs/nifi/processors/RemoteProcessor.java
package org.dcs.nifi.processors;

import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

import org.apache.nifi.annotation.behavior.SideEffectFree;
import org.apache.nifi.annotation.documentation.CapabilityDescription;
import org.apache.nifi.annotation.documentation.Tags;
import org.apache.nifi.annotation.lifecycle.OnScheduled;
import org.apache.nifi.components.PropertyDescriptor;
import org.apache.nifi.flowfile.FlowFile;
import org.apache.nifi.flowfile.attributes.CoreAttributes;
import org.apache.nifi.logging.ProcessorLog;
import org.apache.nifi.processor.AbstractProcessor;
import org.apache.nifi.processor.ProcessContext;
import org.apache.nifi.processor.ProcessSession;
import org.apache.nifi.processor.ProcessorInitializationContext;
import org.apache.nifi.processor.Relationship;
import org.apache.nifi.processor.exception.ProcessException;
import org.apache.nifi.processor.io.OutputStreamCallback;
import org.dcs.api.service.ModuleFactoryService;
import org.dcs.api.service.RESTException;
import org.dcs.remote.RemoteService;


public abstract class RemoteProcessor extends AbstractProcessor {

	private ProcessorLog log = this.getLogger();

	private String flowModuleId;

	private ModuleFactoryService moduleFactoryService;
	private List<PropertyDescriptor> properties;
	private Set<Relationship> relationships;

	protected abstract String getFlowModuleClassName();
	
	@Override
	public void init(final ProcessorInitializationContext context){
		log = this.getLogger();		

		RemoteService.initialize("localhost:2181");
		moduleFactoryService = (ModuleFactoryService) RemoteService.getService(ModuleFactoryService.class);
		flowModuleId = moduleFactoryService.createFlowModule(getFlowModuleClassName());
		
		log.warn("Created flow module " + getFlowModuleClassName() + " with id " + flowModuleId);
		
		Map<String, Properties> propertiesMap = moduleFactoryService.getPropertyDescriptors(flowModuleId);
		properties = ProcessorUtils.generatePropertyDescriptors(propertiesMap);
		
		log.warn("Generated property descriptors ");
		
		Map<String, Properties> relationshipMap = moduleFactoryService.getRelationships(flowModuleId);
		relationships = ProcessorUtils.generateRelationships(relationshipMap);
		
		log.warn("Generated relationships ");
		
	}

	@OnScheduled
  public void onScheduled(final ProcessContext context) {

	}
  
	@Override
	public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
		final Properties valueProperties = ProcessorUtils.generateValueProperties(context.getProperties());
    FlowFile flowFile = session.create();
    flowFile = session.write(flowFile, new OutputStreamCallback() {
        @Override
        public void process(final OutputStream out) throws IOException {
        		byte[] response;
						try {
							response = moduleFactoryService.trigger(flowModuleId, valueProperties);
						} catch (RESTException e) {
							throw new IOException(e);
						}
            out.write(response);
        }
    });

    final Map<String, String> attributes = new HashMap<>();
    attributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
    attributes.put(CoreAttributes.FILENAME.key(), flowFile.getAttribute(CoreAttributes.FILENAME.key()) + ".json");
    flowFile = session.putAllAttributes(flowFile, attributes);

    Relationship successRelationship = ProcessorUtils.getSuccessRelationship(relationships);
    if(successRelationship != null) {
    	session.transfer(flowFile, successRelationship);
    }
                                                                                                                                                                                                                                                                                                                                                                                             
	}

	@Override
	public Set<Relationship> getRelationships(){
		return relationships;
	}

	@Override
	public List<PropertyDescriptor> getSupportedPropertyDescriptors(){
		return properties;
	}

}

File: org.dcs.nifi.processors/src/main/scala/org/dcs/nifi/processors/RemoteProcessor.scala
package org.dcs.nifi.processors

import org.apache.nifi.processor.ProcessContext
import org.apache.nifi.processor.Relationship
import org.apache.nifi.processor.AbstractProcessor
import org.apache.nifi.processor.ProcessorInitializationContext
import org.dcs.api.service.ModuleFactoryService
import org.dcs.remote.ZkRemoteService
import org.dcs.remote.RemoteService
import org.apache.nifi.components.PropertyDescriptor
import org.apache.nifi.annotation.lifecycle.OnScheduled
import org.apache.nifi.processor.ProcessSession
import java.util.Properties
import collection.JavaConversions._
import scala.collection.JavaConverters._
import org.apache.nifi.flowfile.FlowFile
import org.apache.nifi.processor.io.OutputStreamCallback
import java.io.OutputStream
import org.apache.nifi.flowfile.attributes.CoreAttributes
import java.util.{Map  => JavaMap}
import java.util.{Set  => JavaSet}
import java.util.{List => JavaList}
import scala.collection.mutable.{ Map => MutableMap }
import org.slf4j.LoggerFactory
import org.slf4j.Logger

abstract class RemoteProcessor extends AbstractProcessor {

  val logger: Logger = LoggerFactory.getLogger(classOf[RemoteProcessor])

  var flowModuleId: String = _
  var propertyDescriptors: JavaList[PropertyDescriptor] = Nil
  var relationships: JavaSet[Relationship] = _
  var moduleFactoryService: Option[ModuleFactoryService] = None
  
  var remoteService: RemoteService = _

  def flowModuleClassName(): String

  override def init(context: ProcessorInitializationContext) {

    if(remoteService == null) remoteService = ZkRemoteService 
    
    moduleFactoryService = remoteService.loadService[ModuleFactoryService]

    if (moduleFactoryService != None) {

      flowModuleId = moduleFactoryService.get.createFlowModule(flowModuleClassName())

      logger.info("Created flow module " + flowModuleClassName() + " with id " + flowModuleId)

      val propDescMap: JavaMap[String, JavaMap[String, String]] = moduleFactoryService.get.getPropertyDescriptors(flowModuleId)
      propertyDescriptors = ProcessorUtils.generatePropertyDescriptors(propDescMap)

      val relationshipsMap: JavaMap[String, JavaMap[String, String]] = moduleFactoryService.get.getRelationships(flowModuleId)
      relationships = ProcessorUtils.generateRelationships(relationshipsMap)

    } else {
      throw new IllegalStateException("Remote service module of type " + flowModuleClassName + " with module id " + flowModuleId + " not available")
    }

  }

  @OnScheduled
  def onScheduled(context: ProcessContext) {

  }

  override def onTrigger(context: ProcessContext, session: ProcessSession) {
    val valueProperties = ProcessorUtils.valueProperties(context.getProperties())
    var flowFile: FlowFile = session.create();
    flowFile = session.write(flowFile, new OutputStreamCallback() {
      override def process(out: OutputStream) {
        out.write(moduleFactoryService.get.trigger(flowModuleId, valueProperties));
      }
    })

    val attributes = scala.collection.mutable.Map[String, String]()
    attributes(CoreAttributes.MIME_TYPE.key()) = "application/json"
    attributes(CoreAttributes.FILENAME.key()) = flowFile.getAttribute(CoreAttributes.FILENAME.key()) + ".json"
    flowFile = session.putAllAttributes(flowFile, attributes);

    val successRelationship: Option[Relationship] = ProcessorUtils.successRelationship(relationships);
    if (successRelationship != None) {
      session.transfer(flowFile, successRelationship.get);
    }

  }

  override def getRelationships(): JavaSet[Relationship] = {
    return relationships;
  }

  override def getSupportedPropertyDescriptors(): JavaList[PropertyDescriptor] = {
    return propertyDescriptors;
  }

}